connectors/cli.py

```diff
@@ -89,14 +89,14 @@ async def _start_service(config, loop):
     services = [SyncService(config), JobCleanUpService(config)]
     for sig in (signal.SIGINT, signal.SIGTERM):
 
-        async def _shutdown(sig_name):
+        def _shutdown(sig_name):
             logger.info(f"Caught {sig_name}. Graceful shutdown.")
             for service in services:
                 logger.info(f"Shutdown {service.__class__.__name__}...")
-                await service.stop()
+                service.stop()
             return
 
-        loop.add_signal_handler(sig, lambda: asyncio.ensure_future(_shutdown(sig.name)))
+        loop.add_signal_handler(sig, functools.partial(_shutdown, sig.name))
```

Comment:
Can anyone help me understand, why we use `functools.partial` here? We also used it here:

https://github.com/elastic/connectors-python/blob/e6a1e9463f0633748c3cdf7a65021e5950221fb2/connectors/services/sync.py#L139-L141

From what I understand [here](https://docs.python.org/3/library/functools.html#functools.partial), it just wraps the function and freezes some arguments. But I didn't see any additional arguments are used here.

Reply:
two changes:

- the `_shutdown` function is now sync
- lambda is [late binding](https://docs.python-guide.org/writing/gotchas/#late-binding-closures), and `functools.partial` is not 

So if we keep `lambda` we will lookup `sig.name` at execution time. I think I missed this in reviews, this probably does not work as expected. The right one to use is `functools.partial`

Reply:
I see [here](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.add_signal_handler):

>Use [functools.partial()](https://docs.python.org/3/library/functools.html#functools.partial) [to pass keyword arguments](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio-pass-keywords) to callback.

If `_shutdown` doesn't accept any argument, I assume we can do this without `functools.partial`?
```python
loop.add_signal_handler(sig, _shutdown)
``` 

BTW, I checked out your branch and shutdown doesn't work as expected:

<details>
  <summary>Stacktrace</summary>

```  
[FMWK][17:51:33][INFO] Caught SIGINT. Graceful shutdown.
[FMWK][17:51:33][INFO] Shutdown SyncService...
[FMWK][17:51:33][INFO] Shutdown JobCleanUpService...
Exception in callback ConcurrentTasks._callback(result_callback=None)(<Task cancell...s/sync.py:54>>)
handle: <Handle ConcurrentTasks._callback(result_callback=None)(<Task cancell...s/sync.py:54>>)>
Traceback (most recent call last):
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/connectors/services/sync.py", line 95, in _one_sync
    await validate_filtering(
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/connectors/filtering/validation.py", line 29, in validate_filtering
    validation_result = await connector.source_klass(
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/connectors/source.py", line 159, in validate_filtering
    return await FilteringValidator(
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/connectors/filtering/validation.py", line 206, in validate
    filtering_validation_result += await validator.validate(advanced_rules)
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/connectors/sources/mysql.py", line 57, in validate
    return await self._remote_validation(advanced_rules, databases_to_filter)
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/connectors/utils.py", line 380, in func_to_execute
    return await func(*args, **kwargs)
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/connectors/sources/mysql.py", line 65, in _remote_validation
    await self.source.ping()
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/connectors/sources/mysql.py", line 263, in ping
    self.connection_pool = await aiomysql.create_pool(**connection_string)
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/lib/python3.10/site-packages/aiomysql/pool.py", line 29, in _create_pool
    await pool._fill_free_pool(False)
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/lib/python3.10/site-packages/aiomysql/pool.py", line 182, in _fill_free_pool
    conn = await connect(echo=self._echo, loop=self._loop,
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/lib/python3.10/site-packages/aiomysql/connection.py", line 75, in _connect
    await conn._connect()
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/lib/python3.10/site-packages/aiomysql/connection.py", line 540, in _connect
    await self._request_authentication()
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/lib/python3.10/site-packages/aiomysql/connection.py", line 833, in _request_authentication
    auth_packet = await self._read_packet()
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/lib/python3.10/site-packages/aiomysql/connection.py", line 598, in _read_packet
    packet_header = await self._read_bytes(4)
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/lib/python3.10/site-packages/aiomysql/connection.py", line 646, in _read_bytes
    data = await self._reader.readexactly(num_bytes)
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/streams.py", line 708, in readexactly
    await self._wait_for_data('readexactly')
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/streams.py", line 502, in _wait_for_data
    await self._waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/chenhui.wang/src/github.com/elastic/connectors-python/connectors/utils.py", line 306, in _callback
    if task.exception():
asyncio.exceptions.CancelledError
[FMWK][17:51:33][DEBUG] Sleep canceled
[FMWK][17:51:33][INFO] Bye
```
</details>

Reply:
> If _shutdown doesn't accept any argument, I assume we can do this without functools.partial?

Yes, I think that's the case.


> BTW, I checked out your branch and shutdown doesn't work as expected:

What were the steps that you took?

I checked it now and it works well for me.

Reply:
Hmmm I've checked and it indeed happens, but it's not actually connected to the change I've made - one more piece of logic is missing "except CancelledError" before validating filtering and such, I'll address it separately.